api/include/opentelemetry/nostd/variant.h

```diff
@@ -0,0 +1,1279 @@
+// MPark.Variant
+//
+// Copyright Michael Park, 2015-2017
+// Copyright OpenTelemetry Authors, 2020
+//
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file third_party/boost/LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <array>
+#include <exception>
+#include <limits>
+
+#include "opentelemetry/nostd/detail/all.h"
+#include "opentelemetry/nostd/detail/dependent_type.h"
+#include "opentelemetry/nostd/detail/find_index.h"
+#include "opentelemetry/nostd/detail/functional.h"
+#include "opentelemetry/nostd/detail/recursive_union.h"
+#include "opentelemetry/nostd/detail/trait.h"
+#include "opentelemetry/nostd/detail/type_pack_element.h"
+#include "opentelemetry/nostd/detail/variant_alternative.h"
+#include "opentelemetry/nostd/detail/variant_fwd.h"
+#include "opentelemetry/nostd/detail/variant_size.h"
+#include "opentelemetry/nostd/type_traits.h"
+#include "opentelemetry/nostd/utility.h"
+#include "opentelemetry/version.h"
+
+#define AUTO_RETURN(...) \
+  ->decay_t<decltype(__VA_ARGS__)> { return __VA_ARGS__; }
+
+#define AUTO_REFREF_RETURN(...)                                           \
+  ->decltype((__VA_ARGS__))                                               \
+  {                                                                       \
+    static_assert(std::is_reference<decltype((__VA_ARGS__))>::value, ""); \
+    return __VA_ARGS__;                                                   \
+  }
+
+#define DECLTYPE_AUTO_RETURN(...) \
+  ->decltype(__VA_ARGS__) { return __VA_ARGS__; }
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace nostd
+{
+constexpr std::size_t variant_npos = static_cast<std::size_t>(-1);
+
+class bad_variant_access : public std::exception
+{
+public:
+  virtual const char *what() const noexcept override { return "bad_variant_access"; }
+};
+
+[[noreturn]] inline void throw_bad_variant_access()
```

Comment:
This code unconditionally uses **[[noreturn]]** here, which means it does require C++11 as a minimum. So in the other spot you also don't have to check for **codecvt** presence, it's just gonna be there... OR you'd need a compiler-version check here OR not use **[[noreturn]]** attribute at all.

Reply:
C++11 isn't all or nothing.

Older compilers have support for parts of it.

This code should be portable. It's taken from https://github.com/mpark/variant and tested against a range of older compilers
https://github.com/mpark/variant#requirements

Reply:
I meant to say that the **[[noreturn]]** is part of C++11. Same with **codecvt** . If one place of code (here) unconditionally requires a feature that only _officially_ exists in C++11, then we may as well require the other parts of code to freely use C++11 features too without performing any extra checks..

Otherwise we'd try to support older compilers that may have only partial support of that. Then, why not supporting vs2013, and that'd be a problem for heavily templated code and many expressions (like constexpr, static initializers, etc.). We probably just have to say that we only support compilers that have "good" (near-complete) support of C++11 and above.

Reply:
gcc 4.8 is still widely used and support most of C++11, but not codecvt. But why worry about this when it's easy to work around.